python_modules/dagster-test/dagster_test/toys/user_computed_data_versions/external_system.py

```diff
@@ -0,0 +1,155 @@
+import json
+import os
+from dataclasses import asdict, dataclass
+from hashlib import sha256
+from time import sleep
+from typing import AbstractSet, Any, Mapping, Optional, Tuple
+
+import yaml
+
+# ########################
+# ##### ASSET GRAPH INFO
+# ########################
+
+_DEPENDENCIES: Mapping[str, AbstractSet[str]] = {
+    "alpha": set(),
+    "beta": {"alpha"},
+    "epsilon": {"delta"},
+}
+
+_SOURCE_ASSETS: Mapping[str, Any] = {
+    "delta": 100,
+}
+
+
+def _compute_value(key, inputs: Mapping[str, Any]) -> Any:
+    if key == "alpha":
+        return 1
+    elif key == "beta":
+        sleep(10)
+        value = inputs["alpha"] + 1
+        return value
+    elif key == "epsilon":
+        sleep(10)
+        return inputs["delta"] * 5
+
+
+# ########################
+# ##### SYSTEM
+# ########################
+
+
+class ExternalSystem:
+    def __init__(self):
+        self._db = _Database()
+
+    def set_storage_directory(self, storage_directory: str) -> None:
+        self._db.root_path = storage_directory
+
+    def get_code_version(self, key: str) -> str:
+        return "1"
+
+    def get_data_version(self, key: str) -> str:
+        return self._db.get(key).data_version
+
+    def get_dependencies(self, key: str) -> AbstractSet[str]:
+        return _DEPENDENCIES[key]
+
+    def materialize(
+        self, key: str, provenance_json: Optional[str]
+    ) -> Tuple[str, bool]:  # (data_version, is_memoized)
```

Comment:
named tuple instead of comment?

Reply:
Trying to make the interface between the Dagster code and the `ExternalSystem` only use basic collections to more realistically model the data structures used to communicate with an external system (that's why the provenance comes in as json rather than the Dagster `DataProvenance` object.